### PR TITLE
Refactor to fix issues we encountered during SB2018

### PR DESCRIPTION
**Pull request by trickeydan**
Improvements include:
* Support for Pin 13.
* Single character commands to counteract serial buffer issues.
* More understandable names in places ( e.g CommandError -> CommandResponse, commandId -> requestID)

Please note that CI has been removed, pending @RealOrangeOne to rewrite it.

This makes breaking changes to `robotd` and `docs`.

### CODE REVIEW

**Comment by RealOrangeOne** [comment]
(It's unknown whether this actually fixes the issues faced at SB2018)

**Comment by kierdavis** [comment]
The code looks fine to me. I can't vouch for whether it works or not, but you're in a better position to check that than I am :)

What's the reason for removing the CI? The build process doesn't seem to have changed.

It looks like this still does a lot of `String` munging, which is what I think is causing at least some of the weird behaviours we've been seeing (`String`s are heap-allocated, the Arduino has very little heap space, and we don't do any error checking for out-of-memory conditions). However, reducing the verbosity of the protocol and making pin 13 available are definitely steps the right direction.

**Comment by RealOrangeOne** [comment]
> What's the reason for removing the CI

Mostly just so it doesn't blow up. But I think it can be added back now. We should replace it with `arduino-builder` at some point

**Comment by kierdavis** [comment]
@RealOrangeOne Fair enough

File: src/firmware.cpp [hunk 1]
```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
...
+  CommandHandler('L', &led), // Control the debug LED (H/L)
+  CommandHandler('R', &readPin), // Read a digital pin <number>
+  CommandHandler('S', &servo), // Control a servo <num> <width>
+  CommandHandler('U', & ultrasoundRead), // Read an ultrasound distance
+  CommandHandler('V', &version), // Get firmware version
+  CommandHandler('W', &writePin), // Write to or  a GPIO pin <number> <state>
+};
+
+static void dispatch_command(int requestID, const class CommandHandler& handler, const String& argument) {
+  auto err = handler.run(requestID, argument);
```
Comment (RealOrangeOne) [inline src/firmware.cpp]: as it's been renamed to `CommandResponse`, might as well update this to `response`

File: src/firmware.cpp [hunk 2]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
-    digitalWrite(pin, HIGH);
-  } else if (pinStateArg == "low") {
-    pinMode(pin, OUTPUT);
-    digitalWrite(pin, LOW);
-  } else if (pinStateArg == "hi-z") {
-    pinMode(pin, INPUT);
-  } else if (pinStateArg == "pullup") {
-    pinMode(pin, INPUT_PULLUP);
+static CommandResponse led(int requestID, String argument) {
+  if (argument == "H") {
```
Comment (RealOrangeOne) [inline src/firmware.cpp]: This might work better as a `switch`?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+| L       | Control the debug LED       | State {H, L} |                        |
+| R       | Read a digital pin          | Pin Number   |                        |
+| S       | Control a servo             | Servo Number | Width                  |
+| U       | Read an ultrasound distance | Trigger Pin  | Echo Pin               |
+| V       | Get the firmware version    |              |                        |
+| W       | Write to a Pin              | Pin Number   | Pin State {H, L, P, Z} |
+
+## Example Commands
+
+Read the analogue pins: `A`
```
Comment (RealOrangeOne) [inline README.md:25]: Great example! :+1:

File: src/firmware.cpp [hunk 4]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
-static CommandError servo(int commandId, String argument) {
-  String servoArg = pop_option(argument);
-  String widthArg = pop_option(argument);
+// The actual commands
 
-  if (argument.length() || !servoArg.length() || !widthArg.length()) {
-    return COMMAND_ERROR("servo takes exactly two arguments");
-  }
+static void readAnaloguePin(int requestID, const String& name, int pin) {
+  int reading = analogRead(pin);
```
Comment (RealOrangeOne) [inline src/firmware.cpp]: I feel like this temporary variable is unnecessary, might as well pass it to `String` directly

File: src/firmware.cpp [hunk 5]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
 }
 
-static CommandError ultrasound_read(int commandId, String argument) {
+static CommandResponse ultrasoundRead(int requestID, String argument) {
   String triggerPinStr = pop_option(argument);
   String echoPinStr = pop_option(argument);
 
   if (argument.length() || !triggerPinStr.length() || !echoPinStr.length()) {
-    return COMMAND_ERROR("need exactly two arguments: <trigger-pin> <echo-pin>");
+    return COMMAND_ERROR("Bad Arguments");
```
Comment (RealOrangeOne) [inline src/firmware.cpp]: I think the old error message worked better

**Comment by prophile** [comment]
There's a few unrelated changes here; would have been good to factor them out to multiple PRs.

File: src/firmware.cpp [hunk 6]
```diff
@@ -2,51 +2,33 @@
...
-static const float ULTRASOUND_COEFFICIENT = 1e-6 * 343.0 * 0.5 * 1e3;
-
-static const String FIRMWARE_VERSION = "SourceBots PWM/GPIO v0.0.1";
-
-typedef String CommandError;
+static Adafruit_PWMServoDriver SERVOS = Adafruit_PWMServoDriver();
 
-static const CommandError OK = "";
+typedef String CommandResponse;
+static const CommandResponse OK = "";
```
Comment (prophile) [inline src/firmware.cpp:8]: nice

File: src/firmware.cpp (hunk 4 above)
Comment (prophile) [reply src/firmware.cpp]: "reading" is a slightly odd variable name; I'd expect a noun rather than a participle.

File: src/firmware.cpp [hunk 7]
```diff
@@ -62,28 +62,24 @@ static void dispatch_command(int commandId, const class CommandHandler& handler,
   if (err == OK) {
     serialWrite(commandId, '+', "OK");
   } else {
-    serialWrite(commandId, '-', String("Error: ") + err);
+    serialWrite(commandId, '-', "Error: " + err);
   }
 }
 
 static void handle_actual_command(int commandId, const String& cmd) {
+
```
Comment (prophile) [inline src/firmware.cpp]: I think the rest of the file doesn't have this extra line of whitespace.

File: src/firmware.cpp (hunk 2 above)
Comment (prophile) [reply src/firmware.cpp]: `==` is overridden for `String` so I don't think this can be used in a switch.

File: src/firmware.cpp [hunk 8]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
-static CommandError run_help(int commandId, String argument);
+static void serialWrite(int requestID, char lineType, const String& str) {
+    if (requestID != 0) {
+        Serial.write('@');
+        Serial.print(requestID, DEC);
+        Serial.write(' ');
+    }
+
+    Serial.write(lineType);
+    Serial.write(' ');
```
Comment (prophile) [inline src/firmware.cpp]: The indentation here is inconsistent (there's 2 space indentation elsewhere).

File: src/firmware.cpp [hunk 9]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
   return OK;
 }
 
-static CommandError read_pin(int commandId, String argument) {
+static CommandResponse readPin(int requestID, String argument) {
   String pinIDArg = pop_option(argument);
 
   if (argument.length() || !pinIDArg.length()) {
-    return COMMAND_ERROR("need exactly one argument: <pin>");
+    return COMMAND_ERROR("Bad number of arguments");
```
Comment (prophile) [inline src/firmware.cpp:96]: This has less information than the previous error, why the change?

File: src/firmware.cpp (hunk 5 above)
Comment (prophile) [reply src/firmware.cpp]: +1, this has a lot less information than the old error message; if we're looking just for strings to match on then we switch to numeric error codes.

File: src/firmware.cpp [hunk 10]
```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
...
-static void dispatch_command(int commandId, const class CommandHandler& handler, const String& argument) {
-  auto err = handler.run(commandId, argument);
+// Process the commands and execute them.
+
+static const CommandHandler commands[] = {
+  CommandHandler('A', &analogueRead), // Read the analogue pins
+  CommandHandler('L', &led), // Control the debug LED (H/L)
+  CommandHandler('R', &readPin), // Read a digital pin <number>
+  CommandHandler('S', &servo), // Control a servo <num> <width>
+  CommandHandler('U', & ultrasoundRead), // Read an ultrasound distance
```
Comment (prophile) [inline src/firmware.cpp]: There's an extra space after the `&`

File: src/firmware.cpp [hunk 11]
```diff
@@ -340,8 +299,27 @@ static void process_serial() {
...
+  SERVOS.setPWMFreq(50);
+
+  Serial.write("# Booted\n");
+
+}
+
 void loop() {
   while (Serial.available()) {
-    process_serial();
+   process_serial();
```
Comment (prophile) [inline src/firmware.cpp]: ?

File: src/firmware.cpp [hunk 12]
```diff
@@ -60,118 +42,104 @@ static String pop_option(String& argument) {
...
-  }
-  if (width != 0 && (width < 150 || width > 550)) {
-    return COMMAND_ERROR("width must be 0 or between 150 and 550");
-  }
-  SERVOS.setPWM(servo, 0, width);
+static CommandResponse analogueRead(int requestID, String argument) {
+  readAnaloguePin(requestID, "a0", A0);
+  readAnaloguePin(requestID, "a1", A1);
+  readAnaloguePin(requestID, "a2", A2);
+  readAnaloguePin(requestID, "a3", A3);
```
Comment (Tyler-Ward) [inline src/firmware.cpp:70]: Probably worth doing only one of these at a time as anolouge reads take a bit of time.

File: src/firmware.cpp [hunk 13]
```diff
@@ -340,8 +299,27 @@ static void process_serial() {
   serialBuffer += (char)serialInput;
 }
 
+
+void setup() {
+  // Setup the pins.
+  pinMode(LED_BUILTIN, OUTPUT);
+  for (int pin = 2; pin <= 13; pin++) {
```
Comment (prophile) [inline src/firmware.cpp]: The style elsewhere in the file is to use `++pin`

File: src/firmware.cpp [hunk 14]
```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
...
+    serialWrite(requestID, '+', "OK");
   } else {
-    serialWrite(commandId, '-', String("Error: ") + err);
+    serialWrite(requestID, '-', "Error: " + err);
   }
 }
 
-static void handle_actual_command(int commandId, const String& cmd) {
+static void handle_actual_command(int requestID, const String& cmd) {
+
```
Comment (prophile) [inline src/firmware.cpp]: Can we drop the extra spacing here?

**Review by prophile — COMMENTED** [review]
Seems fine, though if we're going for a binary protocol I'd advocate for going the whole hog and re-doing the arguments and return codes also.

File: src/firmware.cpp (hunk 4 above)
Comment (Tyler-Ward) [reply src/firmware.cpp]: It probably reads better to have it as seperate commands, as the intent of teh funtion is seperate from the response generation, the compiler will also create a variable for it anyway,

File: src/firmware.cpp [hunk 15]
```diff
@@ -2,51 +2,33 @@
 
 #include <Adafruit_PWMServoDriver.h>
 
-// Multiplying by this converts round-trip duration in microseconds to distance to object in millimetres.
-static const float ULTRASOUND_COEFFICIENT = 1e-6 * 343.0 * 0.5 * 1e3;
-
-static const String FIRMWARE_VERSION = "SourceBots PWM/GPIO v0.0.1";
-
-typedef String CommandError;
+static Adafruit_PWMServoDriver SERVOS = Adafruit_PWMServoDriver();
```
Comment (Tyler-Ward) [inline src/firmware.cpp:5]: Probably worth adding Doxygen docstrings for the functions and a doc block a the top of the file

File: src/firmware.cpp [hunk 16]
```diff
@@ -2,51 +2,33 @@
...
 
-static const CommandError OK = "";
+typedef String CommandResponse;
+static const CommandResponse OK = "";
 
 #define COMMAND_ERROR(x) ((x))
```
Comment (Tyler-Ward) [inline src/firmware.cpp:10]: This #define just seems to obfiscate the output as it is just a string, might be worth it if the return is a struct or similar and error code is also reported.

File: src/firmware.cpp [hunk 17]
```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
...
+    serialWrite(requestID, '-', String("Error, unknown command: ") + commandIssued);
 }
 
 static void handle_command(const String& cmd) {
+  // A command is prefixed with @ if an identifier is used to prevent race conditions
   if (cmd.startsWith("@")) {
     auto spaceIndex = cmd.indexOf(' ');
-    auto commandId = cmd.substring(1, spaceIndex).toInt();
-    handle_actual_command(commandId, cmd.substring(spaceIndex + 1));
+    auto requestID = cmd.substring(1, spaceIndex).toInt();
```
Comment (RealOrangeOne) [inline src/firmware.cpp:257]: Incoming point of argument: This should probably be `requestId`

File: src/firmware.cpp (hunk 2 above)
Comment (RealOrangeOne) [reply src/firmware.cpp]: I see no reason why it wouldn't, i'd expect the `switch` to use `==` for value comparison anyway

File: src/firmware.cpp (hunk 8 above)
Comment (RealOrangeOne) [reply src/firmware.cpp]: Looks like lots of places have weird indentation. Probably fixed by running auto-indent in any decent editor. The arduino IDE obviously isn't one of them

File: src/firmware.cpp (hunk 1 above)
Comment (trickeydan) [reply src/firmware.cpp]: Fixed in a79124d

File: src/firmware.cpp (hunk 2 above)
Comment (trickeydan) [reply src/firmware.cpp]: This can be changed to a switch by getting the first `char` of the argument. I personally dislike using `String` in `switch` anyway.

Comment (trickeydan) [reply src/firmware.cpp]: Fixed in 6c67c51

File: README.md (hunk 3 above)
Comment (trickeydan) [reply README.md:25]: I have added some non-sarcastic examples in 70dbd21

File: src/firmware.cpp (hunk 4 above)
Comment (trickeydan) [reply src/firmware.cpp]: This variable was leftover from the old code. I have taken @RealOrangeOne's suggestion.

Comment (trickeydan) [reply src/firmware.cpp]: e5da051

File: src/firmware.cpp (hunk 5 above)
Comment (trickeydan) [reply src/firmware.cpp]: I agree that numerical error codes are a better idea. I will switch them back to the previous ones and change to numerical ones if I have time.

Comment (trickeydan) [reply src/firmware.cpp]: See 0cfaddd

File: src/firmware.cpp (hunk 9 above)
Comment (trickeydan) [reply src/firmware.cpp:96]: Fixed in 0cfaddd

File: src/firmware.cpp (hunk 10 above)
Comment (trickeydan) [reply src/firmware.cpp]: Fixed in 885c4e5

File: src/firmware.cpp (hunk 13 above)
Comment (trickeydan) [reply src/firmware.cpp]: Fixed in 885c4e5

File: src/firmware.cpp (hunk 11 above)
Comment (trickeydan) [reply src/firmware.cpp]: ❓

File: src/firmware.cpp (hunk 14 above)
Comment (trickeydan) [reply src/firmware.cpp]: Sure 885c4e5

File: src/firmware.cpp (hunk 12 above)
Comment (trickeydan) [reply src/firmware.cpp:70]: My concern with this is that it will break `robot-api`, I'm not sure if we should be changing our API so much so close to Smallpeice.

File: src/firmware.cpp (hunk 6 above)
Comment (trickeydan) [reply src/firmware.cpp:8]: I agree

File: src/firmware.cpp (hunk 17 above)
Comment (trickeydan) [reply src/firmware.cpp:257]: I'm unsure about what you're pointing out here. Please could you rephrase it?

File: src/firmware.cpp (hunk 12 above)
Comment (trickeydan) [reply src/firmware.cpp:70]: (Have spoken in person and agreed)

File: interface.py [hunk 18]
```diff
@@ -1,82 +0,0 @@
-import serial
```
Comment (PeterJCLaw) [inline interface.py:1]: If at all possible I'd be in favour of keeping this wrapper. While in theory this firmware will mainly be used under [`robotd`](https://github.com/sourcebots/robotd), I've found this wrapper to be very useful for debugging things (both on the arduino and cases where `robotd` wasn't behaving as expected when talking to an arduino).

File: src/firmware.cpp (hunk 9 above)
Comment (PeterJCLaw) [reply src/firmware.cpp:96]: Thanks for fixing most of them, though it looks like this specific case remains.

File: src/firmware.cpp [hunk 19]
```diff
@@ -60,118 +42,110 @@ static String pop_option(String& argument) {
...
 
-static CommandError analogue_read(int commandId, String argument) {
-  read_analogue_pin_to_serial(commandId, "a0", A0);
-  read_analogue_pin_to_serial(commandId, "a1", A1);
-  read_analogue_pin_to_serial(commandId, "a2", A2);
-  read_analogue_pin_to_serial(commandId, "a3", A3);
+  auto width = widthArg.toInt();
+  auto servo = servoArg.toInt();
+  if (servo < 0 || servo > 15) {
+    return COMMAND_ERROR("Servo out of range");
```
Comment (PeterJCLaw) [inline src/firmware.cpp:127]: Can we improve this error message? (Ideally include the valid range and the actual value seen)

File: src/firmware.cpp [hunk 20]
```diff
@@ -340,8 +304,27 @@ static void process_serial() {
...
+    pinMode(pin, INPUT);
+  }
+
+  Serial.begin(9600);
+  Serial.setTimeout(5);
+
+  SERVOS.begin();
+  SERVOS.setPWMFreq(50);
+
+  Serial.write("# Booted\n");
```
Comment (PeterJCLaw) [inline src/firmware.cpp:323]: Previously this was all lowercase; is there anything which relies on an exact string match? (`interface.py` definitely did; I'm not sure whether `robotd` does)

File: interface.py (hunk 18 above)
Comment (trickeydan) [reply interface.py:1]: I'd much rather build this into `robotd` if such an interface is required (reduce code replication). I'd also like to point out that a serial console is pretty useful for this purpose.

File: src/firmware.cpp (hunk 20 above)
Comment (trickeydan) [reply src/firmware.cpp:323]: I did check, it appeared to only be in `interface.py`. `robotd` does not check for this string *at all*, never mind an exact string match.

File: src/firmware.cpp (hunk 17 above)
Comment (RealOrangeOne) [reply src/firmware.cpp:257]: This was a nitpick spelling point. The variable should be called `requestId` rather than `requestID`.

File: src/firmware.cpp (hunk 20 above)
Comment (prophile) [reply src/firmware.cpp:323]: Under the old protocol anything which starts with a '#' is a comment. It is not permissible to assign programmatic semantics to comments.

File: src/firmware.cpp [hunk 21]
```diff
@@ -60,118 +44,110 @@ static String pop_option(String& argument) {
...
   int pin = pinIDArg.toInt();
 
-  if (pin < 2 || pin > 12) {
-    return COMMAND_ERROR("pin must be between 2 and 12");
+  if (pin < 2 || pin > 13) {
+    return COMMAND_ERROR("pin must be between 2 and 13");
   }
 
-  auto state = digitalRead(pin);
+  int state = digitalRead(pin);
```
Comment (prophile) [inline src/firmware.cpp:105]: Pourquoi?

File: interface.py (hunk 18 above)
Comment (prophile) [reply interface.py:1]: Can we put the removal in a separate PR and debate this there?

**Review by prophile — APPROVED** [review]

**Review by kierdavis — APPROVED** [review]
I'd like to see this shipped for the summer school, so I'd be happy if this was merged today. I can see a few opportunities for cleanup (not all of which are related to this PR), so I'd highly recommend revisiting this project after the summer school.
